Add SDN Scalability blog post. (#852)
SovereignCloudStack/website: _i18n/en/_posts/blog/2024-02-05-sdn-scalability.md

---
layout: post
title: "SDN scalability improvements"
author:
  - "Angel Kafazov"
avatar:
  - "akafazov.jpg"
about:
  - "akafazov"
---

# The Data-Center network

<figure class="figure mx-auto d-block" style="width:50%">
  <a href="{% asset "blog/dc-network-architectures.png" @path %}">
    {% asset 'blog/dc-network-architectures.png' class="figure-img w-100" %}
  </a>
</figure>

> A modern data center network typically uses a 2-tier spine-leaf network architectures. A more traditional 3-tier architectures are less popular.

Reviewer: ```suggestion
A modern data center network typically uses a 2-tier spine-leaf network architecture. A more traditional 3-tier architecture is less popular.
```

On top of this physical network, also called underlay, all Software Defined Networking (SDN) features are build on top of. The SDN is the virtual network managed by end users and connects VMs, containers and sometimes physical servers. The SDN is entirely virtualized and must be designed and implemented with the physical topology in mind in order to achieve the desired performance, functionality and scalability requirements.

# Overview of SDN in Openstack

Software Defined Networking (SDN) refers to a networking concept where software-based controllers and APIs are used to direct network traffic rather than the networking hardware. In Openstack, SDN networks are used to create and manage user networks for virtual machines and containers on top of physical network infrastructure. Both types of networks are also referred to as underlay (physical) and overlay (SDN). Larger use cases (hyperscalers) and increased number of containerized workloads put stress on the SDN stack.

> Openstack Neutron is the main component responsible for networking. It heavily leverages open-source technologies like OVN/OVS and also communicates with the underlay network in order to provide higher level network services to users. ML2 is the Neutron plugin which is responsible for communication with the physical network and it can have multiple drivers each supporting particular hardware/vendor devices. The SDN stack in is distributed where some components run on dedicated network nodes, network devices and on the servers as well. 

Reviewer: ```suggestion
OpenStack Neutron is the main component responsible for networking. It heavily leverages open-source technologies like OVN/OVS and also communicates with the underlay network to provide higher-level network services to users. ML2 is the Neutron plugin which is responsible for communication with the physical network and it can have multiple drivers each supporting particular hardware/vendor devices. The SDN stack is distributed where some components run on dedicated network nodes, network devices, and on servers as well.
```

> When Openstack grows its network must scale with it. There are several aspects when we talk about SDN scalability which depends on the actual network architecture. We dive deeper below, but some important points to consider are SDN bottlenecks (OVN control plane, network node resources), support for number of tenants and networks, and actual dataplane performance. In this post we will explorer those challenges and strategies on how to improve SDN scalability in Openstack/SCS.

Reviewer: ```suggestion
When OpenStack grows its network must scale with it. There are several aspects when we talk about SDN scalability which depends on the actual network architecture. We dive deeper below, but some important points to consider are SDN bottlenecks (OVN control plane, network node resources), support for a number of tenants and networks, and actual data-plane performance. In this post, we will explore those challenges and strategies on how to improve SDN scalability in OpenStack/SCS.
```

## SDN stack

> In this section we will take a look at the more software components needed to implement SDN in Openstack.

Reviewer: ```suggestion
In this section we will take a look at the more software components needed to implement SDN in OpenStack.
```

### Neutron

> Neutron is an OpenStack project to provide “network connectivity as a service” between interface devices (e.g., vNICs) managed by other OpenStack services (e.g., nova). It implements the OpenStack Networking API. Neutron is responsible for the centralized management of tenant networks, devices and address allocation. It orchestrates the network, the devices by using plugins and drivers for each particular technology chose in the stack below.

Reviewer: ```suggestion
Neutron is an OpenStack project to provide “network connectivity as a service” between interface devices (e.g., vNICs) managed by other OpenStack services (e.g., nova). It implements the OpenStack Networking API. Neutron is responsible for the centralized management of tenant networks, devices, and address allocation. It orchestrates the network, and the devices by using plugins and drivers for each particular technology chosen in the stack
```

### BGP/EVPN

Border Gateway Protocol (BGP) and Ethernet Virtual Private Network (EVPN) are technologies used in conjunction with Neutron for advanced networking functionalities. BGP is a standardized exterior gateway protocol designed to exchange routing and reachability information among autonomous systems on the internet. EVPN extends BGP to support scalable, multi-tenant layer 2 VPN services. In Neutron, BGP/EVPN is used for efficient routing and segmentation in large-scale cloud environments. It enables more robust and flexible networking solutions by allowing for dynamic route advertisement, enhanced traffic engineering, and seamless integration with existing network infrastructures.

### Tunneling protocols - VXLAN/Geneve

> Virtual Extensible LAN (VXLAN) and Generic Network Virtualization Encapsulation (Geneve) are network overlay protocols used in Openstack and OVN/OVS for tunneling network traffic over existing network infrastructures. VXLAN is widely used for encapsulating Ethernet frames over a UDP tunnel, enabling layer 2 networks to extend over layer 3 networks. Geneve, a newer protocol, offers similar capabilities but with additional flexibility and extensibility. In Openstack, VXLAN/Geneve is critical for creating isolated, multi-tenant networks over a shared physical network infrastructure. This encapsulation allows for scalable network segmentation, providing secure and efficient communication channels within cloud environments.

Reviewer: ```suggestion
Virtual Extensible LAN (VXLAN) and Generic Network Virtualization Encapsulation (Geneve) are network overlay protocols used in OpenStack and OVN/OVS for tunneling network traffic over existing network infrastructures. VXLAN is widely used for encapsulating Ethernet frames over a UDP tunnel, enabling layer 2 networks to extend over layer 3 networks. Geneve, a newer protocol, offers similar capabilities but with additional flexibility and extensibility. In OpenStack, VXLAN/Geneve is critical for creating isolated, multi-tenant networks over a shared physical network infrastructure. This encapsulation allows for scalable network segmentation, providing secure and efficient communication channels within cloud environments.
```

## Physical network

> The physical network layer is crucial as it provides the foundational infrastructure over which all virtualized network functions operate. This layer typically involves various hardware vendors who supply the physical networking equipment such as switches, routers, and other networking hardware. These devices are essential for the actual data transmission and routing in a physical setup. Network Operating Systems (NOS) are also a key component of the physical network layer. NOS is the software running on the network devices, enabling them to perform networking functions and interact with higher-level SDN controllers and applications. In OpenStack, the ML2 (Modular Layer 2) plugin in Neutron plays a vital role in the physical network layer. ML2 provides an abstraction layer that supports multiple networking mechanisms in a pluggable manner, allowing Neutron to interface with various types of networking hardware and technologies. This plugin architecture ensures that Neutron can work with a wide range of physical networking equipment and configurations, thus enabling flexibility and scalability in the physical network infrastructure of an OpenStack-based SDN deployment.

Reviewer: ```suggestion
The physical network layer is crucial as it provides the foundational infrastructure over which all virtualized network functions operate. This layer typically involves various hardware vendors who supply the physical networking equipment such as switches, routers, and other networking hardware. These devices are essential for the actual data transmission and routing in a physical setup. Network Operating Systems (NOS) are also a key component of the physical network layer. NOS is the software running on network devices, enabling them to perform networking functions and interact with higher-level SDN controllers and applications. In OpenStack, the ML2 (Modular Layer 2) plugin in Neutron plays a vital role in the physical network layer. ML2 provides an abstraction layer that supports multiple networking mechanisms in a pluggable manner, allowing Neutron to interface with various types of networking hardware and technologies. This plugin architecture ensures that Neutron can work with a wide range of physical networking equipment and configurations, thus enabling flexibility and scalability in the physical network infrastructure of an OpenStack-based SDN deployment.
```

DPUs are another key component in bare metal hosts, offering similar functionalities to SmartNICs but with additional OS they can be managed independently of the host OS by the service provider. This makes them ideal for Openstack deployments for physical server as SDN and core network features provided by the data center can be managed centrally without impacting the user. 

Incorporating SmartNICs and DPUs into bare metal hosts within an OpenStack environment enhances the network's flexibility, efficiency, and performance. This setup is particularly beneficial for workloads that require high throughput, low latency, and secure network communications. As such, bare metal hosts equipped with these advanced NICs are an essential component of modern SDN architectures, particularly in environments where performance and security are critical.

# Common network designs

> Below we will take a closer look into a common network designs used for Openstack deployments.

Reviewer: Plural ('into common network designs') or singular ('a common network design')?

```suggestion
Below we will take a closer look into a common network design used for OpenStack deployments.
```

    {% asset 'blog/VLANs_network_design.png' class="figure-img w-100" %}
  </a>
</figure>

VLANs is the most simple and common design for very small users that are starting to use Openstack. 

This type of network leverages VLANs between network switches and server for the SDN traffic. User workload like VMs and containers on the server are bound to a VLAN usually with OVS. Neutron orchestrates the process of creation and management of user networks both on the network plane (underlay network) and on the servers. ML2 plugin and drivers are used to talk to the network devices and OVN/OVS control plane on the server.

Users have the option to either pre-provision all VLANs ahead of time, so that when new tenant network appears it can be attached to an existing VLAN without needing to talk to the network devices again. The alternative is to use ML2 to dynamically configure new VLAN before each tenant network is created.

### Pros

- This kind of network is easy and simple to implement. It can be further simplified by pre-provisioning all VLANs on the physical network and potentially eliminating ML2. It is a good approach for test-beds and small deployments.

### Cons

> - Most notably, VLANs for SDN doesn't scale well. First we have a theoretical limit of 4096 VLANs on each network switch. More realistic limit would be about 100 Openstack tenants.

Reviewer: ```suggestion
- Most notably, VLANs for SDN don't scale well. First we have a theoretical limit of 4096 VLANs on each network switch. More realistic limit would be about 100 OpenStack tenants.
```

- The approach is also very fragile. A single VM can brake the whole network, thus exposing a huge blast radius. Also, network switches are notorious for being able to persist VLAN configuration. If a switch goes down, its configuration must be replayed after boot-up which is a process leading to significant downtimes.

> - Another drawback, which we see here is the involvement of the physical network which is needed to support the SDN. For each new user network in Openstack, new VLANs must be created and attached to. Even in pre-provisioned scenario, the VLANs must still exist in the underlay, which eventually becomes a bottleneck.

Reviewer: ```suggestion
- Another drawback, which we see here is the involvement of the physical network which is needed to support the SDN. For each new user network in OpenStack, new VLANs must be created and attached to. Even in pre-provisioned scenario, the VLANs must still exist in the underlay, which eventually becomes a bottleneck.
```

## Network-centric SDN

<figure class="figure mx-auto d-block" style="width:50%">
  <a href="{% asset "blog/Network_centric_network_design.png" @path %}">
    {% asset 'blog/Network_centric_network_design.png' class="figure-img w-100" %}
  </a>
</figure>

This network design employs a more network centric approach by implementing most of the SDN functionalities within the physical network devices. The key point is that some kind of tunneling protocol like VXLAN or Geneve is used to encapsulate user network packets and transport them over the network. In a spine-leaf topology, the leaf or ToR will terminate VXLAN endpoints. For the control plane BGP (EVPN) is used to transmit addressing information between switches.

On the server side, a regular VLAN is provisioned to the ToR switch for each tenant network. Since not all tenant networks are needed on each ToR switch, this design scales better than pure VLANs. Also no VLANs are needed between the switches because the underlay is layer 3.

### Pros

> - This kind of network is more scalable, easily handling up to 1000 Openstack tenants. It is also more stable and resilient to change, because of the smaller blast radius.

Reviewer: ```suggestion
- This kind of network is more scalable, easily handling up to 1000 OpenStack tenants. It is also more stable and resilient to change, because of the smaller blast radius.
```

    {% asset 'blog/Server_centric_network_design.png' class="figure-img w-100" %}
  </a>
</figure>

With the server-centric network design, SDN software features are implemented on the servers and not on the network equipment. Tunneling protocols like VXLAN have terminating endpoints on the servers and the addresses for those endpoints are discovered via control plane protocols like BGP/EVPN. Also encap/decap and crypto processing are done on the server as well.

This kind of design greatly reduces reliance on the network to carry out SDN related tasks. The underlay must be Layer 3 and run BGP in order to enable server-centric SDN. There are no further requirements for the underlay.

Because of the limited involvement of the underlay network, this design scales very well. The downside, however, is the datapath processing in done on the server. Example of such processing is VXLAN encapsulation/decapsulation. If no DPU or SmartNIC exists on the server, this logic runs on the CPU, which can negatively impact its performance. Therefore the use of DPU/SmartNICs is very much needed.

### Pros

- very scalable
- very efficient datapath routing (ingress/egress)
- simple network underlay

> - no network orchestration in Openstack

Reviewer: ```suggestion
- no network orchestration in OpenStack
```

---
layout: post
title: "SDN scalability improvements"
author:
  - "Angel Kafazov"
avatar:
  - "akafazov.jpg"
about:
  - "akafazov"
---

# Overview of SDN network scalability

Software-defined Networking (SDN) is a critical component of modern cloud environments, providing the necessary flexibility and scalability to meet the demands of large-scale network infrastructures. As the size and complexity of cloud deployments continue to grow, the need for scalable and efficient SDN solutions becomes increasingly important. In this post, we explore the challenges and strategies for improving SDN scalability within Sovereign Cloud Stack (SCS) and OpenStack environments.

# The data center network

<figure class="figure mx-auto d-block" style="width:50%">
  <a href="{% asset "blog/dc-network-architectures.png" @path %}">
    {% asset 'blog/dc-network-architectures.png' class="figure-img w-100" %}
  </a>
</figure>

> A modern data center network typically uses a 2-tier spine-leaf network architecture. More traditional 3-tier architectures are becoming less popular.

Author: I don't want to elaborate on 3-tier architecture because we are generally not targeting it. It is getting phased out and not so popular any more.

> In a spine-leaf network, within a rack all servers are connected to a one or two TOR (top of rack) or Leaf switches. All leaf switches are connected to all spine switches in the layer above. There are no direct connections between leaf or spine switches on the same tier. This kind of network facilitates scalability by being simple and much easy to support.

Reviewer 2: ```suggestion
In a spine-leaf network, all servers are connected to one or two TOR (top of rack) or Leaf switches within a rack. All leaf switches are connected to all spine switches in the layer above. There are no direct connections between leaf or spine switches on the same tier. This kind of network facilitates scalability by being simple and easy to support.
```

> On top of this physical network, also called underlay, all Software Defined Networking (SDN) features are build on top of. The SDN is the virtual network managed by end users and connects VMs, containers, and sometimes physical servers. The SDN is entirely virtualized and must be designed and implemented with the physical topology in mind to achieve the desired performance, functionality, and scalability requirements.

Reviewer 2: ```suggestion
On top of this physical network, also called underlay, all Software Defined Networking (SDN) features are built on top of. The SDN is the virtual network managed by end users and connects VMs, containers, and sometimes physical servers. The SDN is entirely virtualized and must be designed and implemented with the physical topology in mind to achieve the desired performance, functionality, and scalability requirements.
```

## Bare metal hosts

Bare metal hosts in the context of OpenStack and SDN play a significant role in providing high-performance, non-virtualized computing resources. These hosts are typically equipped with advanced network interface cards (NICs), such as SmartNICs and Data Processing Units (DPUs). SmartNICs are intelligent network cards that offload processing tasks that would typically be handled by the server CPU. These include functions like traffic shaping, encryption/decryption, and network packet processing. The use of SmartNICs can lead to improved performance and reduced CPU load on the bare metal hosts.

> DPUs are an evolution of SmartNICs and are also very common in clouds providing physical servers to users. The difference between SmartNICs is that DPUs usually run their own OS can be managed independently of the host OS. They are more like computer within a computer. Bare metal servers are directly accessed by the users which creates a constraint on the SDN network, because it should not collide with user workloads. Offloading SDN to the DPU is the best solution for such use cases.

Reviewer 2: ```suggestion
DPUs are an evolution of SmartNICs and are also very common in clouds providing physical servers to users. The difference between SmartNICs is that DPUs usually run their own OS and can be managed independently of the host OS. They are more like computers within a computer. Bare metal servers are directly accessed by the users, which creates a constraint on the SDN network because it should not collide with user workloads. Offloading SDN to the DPU is the best solution for such use cases.
```

Incorporating SmartNICs and DPUs into bare metal hosts within an OpenStack environment enhances the network's flexibility, efficiency, and performance. This setup is particularly beneficial for workloads that require high throughput, low latency, and secure network communications. As such, bare metal hosts equipped with these advanced NICs are an essential component of modern SDN architectures, particularly in environments where performance and security are critical.

# Common network designs

Below we dive deeper into several common network designs used for SCS/OpenStack deployments.

## VLANs

<figure class="figure mx-auto d-block" style="width:50%">
  <a href="{% asset "blog/VLANs_network_design.png" @path %}">
    {% asset 'blog/VLANs_network_design.png' class="figure-img w-100" %}
  </a>
</figure>

> VLANs are the most simple and popular design for small operators which are just starting to use OpenStack. 

Reviewer 2: ```suggestion
VLANs are the simplest and most popular design for small operators that are just starting to use OpenStack. 
```

Operators have the option to either pre-provision all VLANs ahead of time so that when a new tenant network appears it can be attached to an existing VLAN without needing to talk to the network devices again. The alternative is to use ML2 to dynamically configure new VLAN before each tenant network is created.

### Pros

- This kind of network is easy to implement. It can be simplified even further by pre-provisioning all VLANs on the physical network thus eliminating the need for ML2. It is a good approach for test beds and small deployments.

### Cons

- Most notably, VLANs don't scale well. This is due to a theoretical limit of 4096 VLANs on each network switch. A more realistic limit would be about 100 OpenStack tenants for a single cloud network.

> - The approach is also very fragile. A single VM can break the whole network, thus exposing a huge blast radius. Also, network switches are notorious for being able to persist VLAN configuration. If a switch goes down, its configuration must be replayed after boot-up which is a process leading to significant downtimes.

Reviewer 2: ```suggestion
- The approach is also very fragile. A single VM can break the whole network, thus exposing a huge blast radius. Also, network switches are notorious for being able to persist VLAN configuration. If a switch goes down, its configuration must be replayed after boot-up, which is a process that leads to significant downtime.
```

> - Another drawback, which we see is the involvement of the physical network to support the SDN. For each new tenant network in OpenStack, new VLANs must be provisioned. Even in the pre-provisioned scenario, the VLANs must still exist in the underlay, which eventually becomes a bottleneck.

Reviewer 2: ```suggestion
- Another drawback that we see is the involvement of the physical network in supporting the SDN. For each new tenant network in OpenStack, new VLANs must be provisioned. Even in the pre-provisioned scenario, the VLANs must still exist in the underlay, which eventually becomes a bottleneck.
```

> - Another drawback, which we see is the involvement of the physical network to support the SDN. For each new tenant network in OpenStack, new VLANs must be provisioned. Even in the pre-provisioned scenario, the VLANs must still exist in the underlay, which eventually becomes a bottleneck.

Author: Yes, VLANs are provisioned on the physical network.

On the server side, a regular VLAN is provisioned from the ToR switch for each tenant network. Since not all tenant networks are needed on each ToR switch, this design scales better than pure VLANs. Also, no VLANs are used between the switches if the underlay is layer 3.

### Pros

- This kind of network is more scalable, easily handling up to 1000 OpenStack tenants. It is also more stable and resilient to change, because of the smaller blast radius.
- Since all SDN is offloaded to the physical equipment, server CPUs are not doing any encapsulation/decapsulation and are free to serve the users. Heavy data plane processing is done on dedicated network hardware, which has been optimized for this task. Also, no SmartNICs or DPUs are needed on the servers, which is more cost efficient.

### Cons

> - Since this is a network-centric approach, the network is still very much involved in the SDN. Physical device limits shown before like persisting  switch configuration still apply. 

Reviewer 2: ```suggestion
- Since this is a network-centric approach, the network is still very much involved in the SDN. Physical device limits shown before, like persisting switch configuration, still apply. 
```

- Importantly, traffic between tenant networks (east/west and north/south) is routed via the network nodes, which can become a choke point. So while being better for scalability, this design still has limitations for very large deployments.

## Server-centric SDN

<figure class="figure mx-auto d-block" style="width:50%">
  <a href="{% asset "blog/Server_centric_network_design.png" @path %}">
    {% asset 'blog/Server_centric_network_design.png' class="figure-img w-100" %}
  </a>
</figure>

With the server-centric network design, SDN software features are implemented on the servers and not on the network equipment. Tunneling protocols like VXLAN have terminating endpoints on the servers and the addresses for those endpoints are discovered via control plane protocols like BGP/EVPN. Also, encapsulation/decapsulation and crypto processing are done on the server.

This design reduces reliance on the physical network to carry out SDN-related tasks. The only requirement for the underlay is that it must be Layer 3 and run BGP to enable server-centric SDN. 

> Because of the limited involvement of the physcal network, this design scales very well. The downside, however, is that datapath processing is performed on the server. An example of such processing is VXLAN encapsulation/decapsulation. If no DPU or SmartNIC exists on the server, this logic runs on the CPU, which can negatively impact its performance. It is recommended to deploy servers running SmartNICs to help with data-plane performance.

Reviewer 2: ```suggestion
Because of the limited involvement of the physical network, this design scales very well. The downside, however, is that datapath processing is performed on the server. An example of such processing is VXLAN encapsulation/decapsulation. If no DPU or SmartNIC exists on the server, this logic runs on the CPU, which can negatively impact its performance. It is recommended to deploy servers running SmartNICs to help with data-plane performance.
```

### Pros

- very scalable
- very efficient datapath routing (ingress/egress)
- simple network underlay
- no network orchestration in OpenStack

### Cons

- complex server
- might require SmartNICs and DPUs to offload server CPU
- must run routing protocols in the server

## Hybrid - SDN on both servers and switches

<figure class="figure mx-auto d-block" style="width:50%">
  <a href="{% asset "blog/Hybrid_network_design.png" @path %}">
    {% asset 'blog/Hybrid_network_design.png' class="figure-img w-100" %}
  </a>
</figure>

> A variation of the server-centric design, the hybrid network supports bare metal nodes, which are allocated to single users. Since the user has full access to the OS of the server, the SDN functionality cannot be deployed alongside applications, since SDN is managed by the service provider. We have two valid options for deploying SDN components in such network:

Reviewer 2: ```suggestion
A variation of the server-centric design, the hybrid network supports bare metal nodes, which are allocated to single users. Since the user has full access to the OS of the server, the SDN functionality cannot be deployed alongside applications, since SDN is managed by the service provider. We have two valid options for deploying SDN components in such a network:
```

- dedicated customer switch, attached to a single bare metal node
- a DPU card attached to the bare metal node

> SmartNIC is not a good solution, because it is more tightly coupled to the host node and exposed to the end user. DPUs and switches can be  independently managed by the provider. This separation helps avoid collision between network infrastructure and user workloads.

Reviewer 2: ```suggestion
SmartNIC is not a good solution because it is more tightly coupled to the host node and exposed to the end user. DPUs and switches can be independently managed by the provider. This separation helps avoid collision between network infrastructure and user workloads.
```

On the other hand, DPUs are ideally suited for environments where bare metal servers are directly accessed by end-users. Representing the next evolutionary step beyond SmartNICs, DPUs offer the distinct advantage of being managed independently by the network operator, separate from user control. Within SCS, we provide the essential infrastructure and configuration capabilities, enabling users to effortlessly set up and manage SmartNIC and DPU features on their networks.

## Network design - VXLAN on Switches

When SDN protocols are implemented directly on network switches, our approach involves exploring the use of SONiC alongside enhancements in tooling and automation to scale effectively.

### SONiC

Our strategy includes bolstering support for SONiC, a powerful open-source network operating system that enables network scalability and flexibility.

### Tooling - Automating SONIC rollout and configuration

> We are committed to enhancing tooling, configuration, and documentation to facilitate the adoption of this network architecture in SCS. This includes improvements in OSISM and Kolla Ansible, along with the integration of Netbox. Netbox serves as the source of truth, allowing for the generation of initial roll-out configurations for OSISM and Kolla Ansible. Additionally, we aim to integrate observability and monitoring for SONiC-based network equipment into SCS, leveraging SONiC's existing observability features.

Reviewer 2: It would be nice to add a link to Netbox website.
